### PR TITLE
Full-Time Schedules

### PR DESCRIPTION
**Pull request by k8culver**
The following has been added/removed:
- Employees per shift is now a per day setting (currently called Forecast but let me know if you have a better name)
- Added the concept of full-time employees
- I've removed the exactly 1 manager per shift setting and have assumed at least 1 manager per shift
- I've moved the random number seed to the configs file instead of having it in the UI

Curious to get feedback on collapsing some of the settings under "Advanced Settings" now that I'm thinking about it it seems like there might be some infeasible runs because they don't notice the settings that are hidden.

Closes: https://github.com/dwave-examples/employee-scheduling/issues/24

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,353 +1,90 @@
...
-app = Dash(
+app = dash.Dash(
     __name__,
     meta_tags=[{"name": "viewport", "content": "width=device-width, initial-scale=1"}],
     prevent_initial_callbacks="initial_duplicate",
     background_callback_manager=background_callback_manager,
 )
 app.title = APP_TITLE
 
+app.config.suppress_callback_exceptions = True
```
Comment (thisac) [inline app.py:52]: Why is this needed?

File: demo_callbacks.py [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+    }
+    num_full_time = min(num_full_time, new_full_time_max)
+
+    df = utils.build_random_sched(num_employees, num_full_time)
+
+    init_availability_table = utils.display_availability(df)
+
+    # Prepare forecast defaults
+    df_to_count = df.iloc[:num_full_time, :]
+    count = df_to_count.applymap(lambda cell: cell.count(REQUESTED_SHIFT_ICON)).sum()[1:].to_dict()
```
Comment (thisac) [inline demo_callbacks.py:199]: `df.applymap` has been deprecated in favour of `df.map`.
```suggestion
    count = df_to_count.map(lambda cell: cell.count(REQUESTED_SHIFT_ICON)).sum()[1:].to_dict()
```

File: demo_callbacks.py [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        isolated_days_allowed,
+        consecutive_shifts + 1,
+        num_full_time,
+    )
+
+    feasible_sampleset, errors = employee_scheduling.run_cqm(cqm)
+    sample = feasible_sampleset.first.sample
+
+    sched = utils.build_schedule_from_sample(sample, employees)
+    scheduled_count = sched.applymap(lambda cell: UNAVAILABLE_ICON not in cell).sum()[1:].to_dict()
```
Comment (thisac) [inline demo_callbacks.py:340]:
```suggestion
    scheduled_count = sched.map(lambda cell: UNAVAILABLE_ICON not in cell).sum()[1:].to_dict()
```

File: src/employee_scheduling.py [hunk 4]
```diff
@@ -13,34 +13,34 @@
...
     max_shifts,
-    shift_min,
-    shift_max,
-    requires_manager,
+    shift_forecast,
     allow_isolated_days_off,
     max_consecutive_shifts,
+    num_full_time,
 ):
     """Builds the ConstrainedQuadraticModel for the given scenario."""
```
Comment (thisac) [inline src/employee_scheduling.py:32]: This should have a complete docstring with args, etc.

**Review by thisac — APPROVED** [review]
LGTM, apart from a few minor comments. Works and looks great!

File: app.py (hunk 1 above)
Comment (k8culver) [reply app.py:52]: It's an aspect of the template that disabled Dash's default check of scanning every callback Input, Output, State, etc to make sure the Dash object associated with that id exists in the layout. We don't want this usually because there's cases when we add something to the layout and then have a callback associated with it and if this wasn't disabled it would error because that object doesn't exist on first load.
